// fhw-web/dist/public/config/default-config.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultConfig = void 0;
const languages_1 = require("./languages");
const frontmatter_1 = require("../frontmatter");
const logging_types_1 = require("./logging-types");
const routing_config_1 = require("./routing-config");
exports.DefaultConfig = {
    rootPath: process.cwd(),
    language: languages_1.Languages.DE,
    loggingActive: [logging_types_1.LoggingTypes.INFO, logging_types_1.LoggingTypes.DATA, logging_types_1.LoggingTypes.WARN, logging_types_1.LoggingTypes.ERROR],
    server: {
        host: 'localhost',
        port: 8080,
        logging: true
    },
    routing: {
        magic: false,
        fileName: 'routes',
        fileExtension: routing_config_1.RoutingFileExtensions.JSON,
        reloadOnEveryRequest: true,
        logging: true
    },
    templating: {
        validation: true,
        paths: {
            pages: 'pages',
            templates: 'templates',
            helpers: 'helpers',
            controller: 'controller'
        },
        allowedExtensions: ['html', 'hbs'],
        frontmatterFormat: frontmatter_1.FrontmatterTypes.JSON,
        helpers: {
            reloadOnEveryRequest: true,
        },
        logging: true
    },
    sessions: {
        active: false,
        path: 'sessions',
        logging: true
    },
    database: {
        globalData: {
            active: true,
            reloadOnEveryRequest: true,
            pathToFile: './global.json',
            format: frontmatter_1.FrontmatterTypes.JSON,
            logging: true
        },
        fileData: {
            active: false,
            reloadOnEveryRequest: true,
            path: './data',
            format: frontmatter_1.FrontmatterTypes.JSON,
            logging: true
        },
        sqliteData: {
            active: false,
            reloadOnEveryRequest: true,
            pathToFile: './data/sqlite.db',
            logging: true
        }
    }
};
